### ISSUE TITLE
redirecting >11 times causes "MaxListenersExceededWarning"

### ISSUE THREAD

**Issue by KhafraDev** [nodejs/undici#1711]
## Bug Description

Each redirect adds one `terminated` listener to fetchParams and it's never removed.

## Reproducible By

```mjs
import { once } from 'events'
import { createServer } from 'http'
import { fetch } from 'undici'

let redirects = 0

const server = createServer((req, res) => {
  if (redirects === 15) {
    res.end('Okay goodbye')
    return
  }

  res.writeHead(302, {
    Location: `/${redirects++}`
  })
  res.end()
}).listen(0)

await once(server, 'listening')

const url = `http://localhost:${server.address().port}`

fetch(url, { redirect: 'follow' })
```

## Logs & Screenshots

```
MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 terminated listeners added to [Fetch]. Use emitter.setMaxListeners() to increase limit
```

**Comment by jonseymour** [comment]
Right, but if you have to lift the warning so high that it masks actual leaks you cease to get any benefit from the warning since if there is an actual leak in one of these event sources the warning never fires or if does fires once due a spurious condition then, when an actual leak occurs later,  the code fails to emit a second warning because it has already emitted a warning.

Maybe the most pragmatic thing would be to place a rate limit on the number of warnings, but ensure they are issued periodically along with the current number of listeners. That way, an actual leak could be observed (by observing a steady increase in the watermark) without necessarily polluting the logs with too many spurious warnings.

**Comment by Uzlopak** [comment]
The problem is solved by @KhafraDev. You can open an issue in node core and talk there about more strategies to improve this. 

From the undici perspective the bug will be gone when #2823 is merged.

**Comment by jonseymour** [comment]
Ok, good to see the root cause is being addressed now, thanks.

**Comment by daveyarwood** [comment]
Another note: I did try increasing the default max event listeners to a high number:

```javascript
EventEmitter.defaultMaxListeners = 5000
```

And I'm still seeing the warnings come up. Granted, my service is doing a lot of work in parallel (many promises being fired off), so I'm not sure if the problem might be user error. Are there best practices about using this library in many promises concurrently?

**Comment by daveyarwood** [comment]
For context, I am using undici transitively via Elasticsearch.

**Comment by KhafraDev** [comment]
It doesn't look like elasticsearch is using fetch, which is what #2823 fixed the issue for. I would recommend opening up a new issue for this and, if possible, provide a minimal reproducible example.
